tensorboard/webapp/widgets/line_chart_v2/lib/renderer/renderer_types.ts

```diff
@@ -0,0 +1,39 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+import {Paths, Rect} from '../types';
+
+export interface IRenderer {
+  /**
+   * Certain renderer requires DOM dimensions for correct density and operations. The
+   * method is invoked when container is resized.
+   *
+   * @param domRect Container dimensions
+   */
+  onResize(domRect: Rect): void;
+
+  drawLine(cacheId: string, paths: Paths, spec: LineSpec): void;
+
+  flush(): void;
+
+  renderGroup(groupName: string, renderBlock: () => void): void;
```

Comment:
Follow-up question: I’m getting the impression that this interface
expects its clients to call `drawLine` with the same cache ID many times
in the same render group. Why would they do that? If the render group
lives not longer than a frame (just a guess), and the cache ID
corresponds to one line segment on screen (just a guess), why would a
client draw a given line segment more than once per frame? and thus why
cache at all instead of just painting immediately?